src/core/api/plug_scanners.go

```diff
@@ -0,0 +1,350 @@
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package api
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/goharbor/harbor/src/pkg/plug/scanner/api"
+	"github.com/goharbor/harbor/src/pkg/plug/scanner/dao/scanner"
+	"github.com/goharbor/harbor/src/pkg/q"
+	"github.com/pkg/errors"
+)
+
+// ScannerAPI provides the API for managing the plugin scanners
+type ScannerAPI struct {
+	// The base controller to provide common utilities
+	BaseController
+
+	// Controller for the plug scanners
+	c api.Controller
+}
+
+// Prepare sth. for the subsequent actions
+func (sa *ScannerAPI) Prepare() {
+	// Call super prepare method
+	sa.BaseController.Prepare()
+
+	// Check access permissions
+	if !sa.SecurityCtx.IsAuthenticated() {
+		sa.SendUnAuthorizedError(errors.New("UnAuthorized"))
+		return
+	}
+
+	if !sa.SecurityCtx.IsSysAdmin() {
+		sa.SendForbiddenError(errors.New(sa.SecurityCtx.GetUsername()))
+		return
+	}
+
+	// Use the default controller
+	sa.c = api.DefaultController
+}
+
+// Get the specified scanner
+func (sa *ScannerAPI) Get() {
+	if r := sa.get(); r != nil {
+		// Response to the client
+		sa.Data["json"] = r
+		sa.ServeJSON()
+	}
+}
+
+// List all the scanners
+func (sa *ScannerAPI) List() {
+	p, pz, err := sa.GetPaginationParams()
+	if err != nil {
+		sa.SendBadRequestError(errors.Wrap(err, "scanner API: list all"))
+		return
+	}
+
+	query := &q.Query{
+		PageSize:   pz,
+		PageNumber: p,
+	}
+
+	// Get query key words
+	kws := make(map[string]string)
+	kw := sa.GetString("q")
+	if len(kw) > 0 {
+		ws := strings.Split(kw, ",")
+		for _, w := range ws {
+			tw := strings.TrimSpace(w)
+			kv := strings.Split(tw, "=")
+			if len(kv) == 2 {
+				// valid key and value for the querying keyword
+				kws[kv[0]] = kv[1]
+			}
+		}
+	}
+
+	if len(kws) > 0 {
+		query.Keywords = kws
+	}
+
+	all, err := sa.c.ListRegistrations(query)
+	if err != nil {
+		sa.SendInternalServerError(errors.Wrap(err, "scanner API: list all"))
+		return
+	}
+
+	// Response to the client
+	sa.Data["json"] = all
+	sa.ServeJSON()
+}
+
+// Create a new scanner
+func (sa *ScannerAPI) Create() {
+	r := &scanner.Registration{}
+
+	if err := sa.DecodeJSONReq(r); err != nil {
+		sa.SendBadRequestError(errors.Wrap(err, "scanner API: create"))
+		return
+	}
+
+	if err := r.Validate(false); err != nil {
+		sa.SendBadRequestError(errors.Wrap(err, "scanner API: create"))
+		return
+	}
+
+	// Explicitly check if conflict
+	if !sa.checkDuplicated("name", r.Name) ||
+		!sa.checkDuplicated("url", r.URL) {
+		return
+	}
+
+	uid, err := sa.c.CreateRegistration(r)
```

Comment:
Maybe there will be more than one default scanners when call this API with the true value of `isDefault`.

Reply:
Correct, add guard code here. All the newly created registration should be none default ones.

Reply:
added